helix-term/src/commands.rs

```diff
@@ -4903,3 +4910,311 @@ fn replay_macro(cx: &mut Context) {
+        let cursor_pos = next.cursor(text);
+        let row = visual_coords_at_pos(doc.text().slice(..), cursor_pos, doc.tab_width()).row;
+        if row < view.offset.row {
+            break;
+        }
+        if !view.is_cursor_in_view(cursor_pos, doc, 0) {
+            if cursor_pos == 0 {
+                break;
+            }
+            continue;
+        }
+        if bck_jump_locations
+            .last()
+            .map(|(pos, _)| pos - cursor_pos <= 1)
+            .unwrap_or(false)
+        {
+            continue;
+        }
+        bck_jump_locations.push((cursor_pos, next.anchor));
+        if cursor_pos == 0 {
+            break;
+        }
+    }
+
+    jump_mode_impl(cx, [fwd_jump_locations, bck_jump_locations]);
+}
+
+fn jump_mode_search(cx: &mut Context) {
+    jump_mode_search_impl(cx, false);
+}
+
+fn extend_jump_mode_search(cx: &mut Context) {
+    jump_mode_search_impl(cx, true);
+}
+
+fn jump_mode_search_impl(cx: &mut Context, extend: bool) {
+    cx.on_next_key(move |cx, event| {
+        let c = match event.char() {
+            Some(c) => c,
+            _ => return,
+        };
+
+        let (view, doc) = current!(cx.editor);
+
+        let text = doc.text().slice(..);
+        let (cursor, anchor) = {
+            let range = doc.selection(view.id).primary();
+            (range.cursor(text), range.anchor)
+        };
+
+        let mut fwd_jump_locations = Vec::new();
+        for n in 1.. {
+            let next = search::find_nth_next(text, c, cursor + 1, n);
+            match next {
+                Some(pos) => {
+                    let row = visual_coords_at_pos(doc.text().slice(..), pos, doc.tab_width()).row;
+                    if row >= view.offset.row + view.inner_area().height as usize {
+                        break;
+                    }
+                    if !view.is_cursor_in_view(pos, doc, 0) {
+                        continue;
+                    }
+                    fwd_jump_locations.push((pos, if extend { anchor } else { pos }));
+                }
+                _ => break,
+            }
+        }
+        let mut bck_jump_locations = Vec::new();
+        for n in 1.. {
+            let next = search::find_nth_prev(text, c, cursor, n);
+            match next {
+                Some(pos) => {
+                    let row = visual_coords_at_pos(doc.text().slice(..), pos, doc.tab_width()).row;
+                    if row < view.offset.row {
+                        break;
+                    }
+                    if !view.is_cursor_in_view(pos, doc, 0) {
+                        continue;
+                    }
+                    bck_jump_locations.push((pos, if extend { anchor } else { pos }));
+                }
+                _ => break,
+            }
+        }
+
+        jump_mode_impl(cx, [fwd_jump_locations, bck_jump_locations]);
+    });
+}
+
+fn jump_mode_impl(cx: &mut Context, [fwd_jumps, bck_jumps]: [Vec<(usize, usize)>; 2]) {
```

Comment:
I think the array is awkward here since it isn't iterated. I would prefer these be separate arguments

Also this is a nitpick but I think these would be clearer as `forward_jumps` and 
`backward_jumps`. The abbreviation doesn't save that many characters and reduces readability

Reply:
I originally had the argument as a tuple, but clippy complained about the type being too complex, so I made an array. I never actually considered making them separate arguments.